DOM - Document support for bulk CRUD operations of DomInstances (RN 37891) (#2339)
SkylineCommunications/dataminer-docs: user-guide/Advanced_Modules/DOM/DOM_examples/DOM_BulkProcessing_Example.md

Reviewer: General comments:
- Try to use present tense as much as possible
- Avoid contractions (there's -> there is)
- Always mention the MR with the version

For the `DomInstances` that failed their name and the issue will be logged in this example.

  ```csharp
  // Retrieve the DomInstances that need processing
  var domInstances = domHelper.DomInstances.Read(filter);

  // Update the field value on those DomInstances
  foreach(var domInstance in domInstances)
  {
      domInstance.AddOrUpdateFieldValue<int>(sectionDefinitionId, fieldDescriptorId, newType);
  }

  // Save them to the DB.
  var updateResult = domHelper.DomInstances.CreateOrUpdate(domInstances);

  // Log what items were successfully removed
  Log($"The new type was successfully set on {updateResult.SuccessfulItems.Count}");

  // Log the some updates went wrong. Per DomInstance check how what issue occurred.
  var incorrectInstances = updateResult.TraceDataPerItem.Where(x => !x.Value.HasSucceeded()).ToList();
  if (incorrectInstances.Count > 0)
  {
      var issues = incorrectInstances.Select(x => $"{domInstances.FirstOrDefault(y => x.Equals(y.ID))?.Name}: {x.Value}");

>       Log($"The type couldn't be updated for {incorrectInstances.Count} items: {string.Join(Environment.NewLine, issues)}");

Reviewer: ```suggestion
      Log($"The type could not be updated for {incorrectInstances.Count} items: {string.Join(Environment.NewLine, issues)}");
```

  }

>   ```

Reviewer 2: ```suggestion
```
```

## Create multiple DomInstances

> In this example two `DomInstances` get created. The second `DomInstance` will be created using a `FieldDescriptor` that no longer is supported for initial state it gets created in.

Reviewer: ```suggestion
In this example, two `DomInstances` get created. The second `DomInstance` will be created using a `FieldDescriptor` that no longer is supported for the initial state it gets created in.
```

> In this example the `TraceData` will be logged. Depending on the issue that occurs, the ID of the `DomInstance` might be included in the error.

Reviewer: ```suggestion
In this example, the `TraceData` will be logged. Depending on the issue that occurs, the ID of the `DomInstance` might be included in the error.
```

  ```csharp
  // Instantiate a new DomInstance & assign 'Hello' as a value for a correct FieldDescriptor.
  var domInstanceOne = new DomInstance() { DomDefinitionId = basicDomDefinitionId };
  domInstanceOne.AddOrUpdateFieldValue<string>(sectionDefinitionId, fieldDescriptorId, "Hello");

  // Instantiate a new DomInstance & assign 'World' as a value for a FieldDescriptor no longer supported for the initial state.
  var domInstanceTwo = new DomInstance() { DomDefinitionId = advancedDomDefinitionId };
  domInstanceTwo.AddOrUpdateFieldValue<string>(sectionDefinitionId, incorrectFieldDescriptorId, "World");

  // Save them to the DB
  var domInstances = new List<DomInstance> { domInstanceOne, domInstanceTwo };
  domHelper.DomInstances.CreateOrUpdate(domInstances);

  // Log if an issue occurred
  var traceData = domHelper.DomInstances.GetTraceDataLastCall();
  if (!traceData.HasSucceeded())
  {

>       Log($"Some of the instances couldn't be created: {traceData}");

Reviewer: ```suggestion
      Log($"Some of the instances could not be created: {traceData}");
```

  }
  ```

## Remove multiple DomInstances

> In this example some `DomInstances` get filter out to be removed.

Reviewer: ```suggestion
In this example, some `DomInstances` get filtered out to be removed.
```

> If, for example, the user executing this example isn't allowed to remove them, it will be logged for each `DomInstance`.

Reviewer: ```suggestion
If, for example, the user executing this example is not allowed to remove them, it will be logged for each `DomInstance`.
```

> In case the `Read` or `Delete` call was unsuccessful because of a database issue, a `CrudFailedException` will be thrown. Details of the issue will available in the `TraceData` of that exception.

Reviewer: ```suggestion
In case the `Read` or `Delete` call is unsuccessful because of a database issue, a `CrudFailedException` will be thrown. Details of the issue will be available in the `TraceData` of that exception.
```

  ```csharp
  try
  {
      // Retrieve the DomInstances that need to get deleted
      var domInstances = domHelper.DomInstances.Read(filter);

>       // Remove them to the DB.

Reviewer: ```suggestion
      // Remove them from the DB.
```

      var deleteResult = domHelper.DomInstances.Delete(domInstances);

>       // Log what items were successfully removed

Reviewer: ```suggestion
      // Log what items are successfully removed
```

      Log($"Successfully removed {deleteResult.SuccessfulIds.Count} items");

>       // Log the some updates went wrong. Per DomInstance check how what issue occurred.

Reviewer: ```suggestion
      // Log that some updates went wrong. Per DomInstance check how the issue occurred.
```

  var domInstanceOne = new DomInstance() { DomDefinitionId = domDefinitionId };
  domInstanceOne.AddOrUpdateFieldValue<string>(sectionDefinitionId, fieldDescriptorId, "Hello");

  // Instantiate a new DomInstance & assign 'World' as a value.
  var domInstanceTwo = new DomInstance() { DomDefinitionId = domDefinitionId };
  domInstanceOne.AddOrUpdateFieldValue<string>(sectionDefinitionId, fieldDescriptorId, "World");

  // Save them to the DB.
  var domInstances = new List<DomInstance> { domInstanceOne, domInstanceTwo };
  domHelper.DomInstances.CreateOrUpdate(domInstances);
  ```

## Update a value for multiple DomInstances

In the following example a field gets updated on multiple `DomInstances`.

  ```csharp
  // Retrieve the DomInstances that need processing.
  var domInstances = domHelper.DomInstances.Read(filter);

  // Update the field value on those DomInstances.
  foreach(var domInstance in domInstances)
  {
      domInstance.AddOrUpdateFieldValue<string>(sectionDefinitionId, fieldDescriptorId, newValue);
  }

  // Save them to the DB.
  domHelper.DomInstances.CreateOrUpdate(domInstances);
  ```

### Checking issues

> For some of the `DomInstances` the update might not succeed. In this example the number of `DomInstances` that fails is logged, together with the issues that occurred. Next the number of `DomInstances` that succeeds gets logged.

Reviewer 2: ```suggestion
For some of the `DomInstances` the create or update might not succeed. In this example the number of `DomInstances` that fails is logged, together with the issues that occurred. Next the number of `DomInstances` that succeeds gets logged.
```

  var updateResult = domHelper.DomInstances.CreateOrUpdate(domInstances);

  // Check if some of the updates are not successful.
  if (updateResult.HasFailures())
  {
      // Log the number of DomInstances that was not updated.
      // Also log the `TraceData` for all failing `DomInstances`. The `TraceData` contains all errors and warnings.
      Log($"Could not perform the update for {updateResult.UnsuccessfulIds.Count} items: {updateResult.GetTraceData()}");
  }

  // Log what items were successfully removed.
  Log($"Could perform the update successfully for {updateResult.SuccessfulItems.Count} items");
  ```

Next to a summarized logging using `GetTraceData()`, `TraceDataPerItem` can be used to check for errors and warnings per `DomInstanceId`.

When multiple `DomInstances` are created or deleted, the issues can be logged the same way.

## Remove multiple DomInstances

In this example, some `DomInstances` get filtered out to be removed.

  ```csharp
  // Retrieve the DomInstances that need to get deleted.
  var domInstances = domHelper.DomInstances.Read(filter);

  // Remove them from the DB.
  var deleteResult = domHelper.DomInstances.Delete(domInstances);

>   ```

Reviewer 2: Maybe add something here saying that, on the delete result you can also check what items has succeeded or failed just like shown with the CreateOrUpdate. Only difference here is that you do not have the 'items' collection.

## Unexpected issue

In case a `CreateOrUpdate` or a `Delete` call are unsuccessful, for instance because of a database issue, by default a `CrudFailedException` will be thrown. Details of the issue will be available in the `TraceData` of that exception.

In the following example, some `DomInstances` need to be deleted:

```csharp
try
{
  // Remove them from the DB.
  var deleteResult = domHelper.DomInstances.Delete(domInstances);

  // Log what items are successfully removed.
  Log($"Successfully removed {deleteResult.SuccessfulIds.Count} items");
}
catch (CrudFailedException ex)
{
  // Log the issue that occurs.
  Log($"Cannot perform the removal, because of the following issue: {ex.TraceData}");
}
```

It is also possible to disable this and check if something went wrong yourself by requesting the `TraceData` object of the last call.

```csharp
// Disable the exceptions.
domHelper.DomInstances.ThrowExceptionsOnErrorData = false;

// Remove them from the DB.

> var deleteResult = domHelper.DomInstances.DomInstances.Delete(domInstances);

Reviewer 3: ```suggestion
var deleteResult = domHelper.DomInstances.Delete(domInstances);
```

> // Check if a result is available.

Reviewer 3: Can this block be removed (lines 120-126)? (discussed with TGH that it would be clearer to focus on calling the GetTraceDataLastCall like we do for other methods)

Author: I have my doubts about this one. Primary source of checking what when right or wrong should be *deleteResult*. It should never be null, but because *ThrowExceptionsOnErrorData*, it could be in some rare cases.
Maybe the following would make that clearer:

```cs
if (deleteResult == null)
{
  // Get the TraceData and check if the last call succeeded.
  var traceData = domHelper.DomInstances.GetTraceDataLastCall();

  // Log the issue that occurs.
  Log($"An unexpected issue occurred during the removal, because of the following issue: {traceData}");
  return;
}

// Log what items are successfully removed.
Log($"Successfully removed {deleteResult.SuccessfulIds.Count} items");
```

Author: As discussed I've implemented the above suggestion and also included the *HasFailures* path.

fyi. I didn't add that in the example initially, since currently on DomInstance level there's no proxy that will cause an error during a delete. But such a proxy could get added in the future.
